src/pudl/extract/extractor.py

```diff
@@ -0,0 +1,375 @@
+
+        for page in self._metadata.get_all_pages():
+            if page in self.BLACKLISTED_PAGES:
+                logger.debug(f"Skipping blacklisted page {page}.")
+                continue
+            current_page_dfs = [
+                pd.DataFrame(),
+            ]
+            for partition in pudl.helpers.iterate_multivalue_dict(**partitions):
+                # we are going to skip
+                if self.source_filename(page, **partition) == "-1":
+                    logger.debug(f"No page for {self._dataset_name} {page} {partition}")
+                    continue
+                logger.debug(
+                    f"Loading dataframe for {self._dataset_name} {page} {partition}"
+                )
+                df = self.load_source(page, **partition)
+                df = pudl.helpers.simplify_columns(df)
+                df = self.process_raw(df, page, **partition)
+                df = self.process_renamed(df, page, **partition)
+                self.validate(df, page, **partition)
+                current_page_dfs.append(df)
+
+            all_page_dfs[page] = self.combine(current_page_dfs, page)
+        return all_page_dfs
+
+
+@op
+def concat_pages(paged_dfs: list[dict[str, pd.DataFrame]]) -> dict[str, pd.DataFrame]:
+    """Concatenate similar pages of data from different years into single dataframes.
+
+    Transform a list of dictionaries of dataframes into a single dictionary of
+    dataframes, where each dataframe is the concatenation of dataframes with identical
+    keys from the input list.
+
+    Args:
+        paged_dfs: A list of dictionaries whose keys are page names, and values are
+            extracted DataFrames. Each element of the list corresponds to a single
+            year of the dataset being extracted.
+
+    Returns:
+        A dictionary of DataFrames keyed by page name, where the DataFrame contains that
+        page's data from all extracted years concatenated together.
+    """
+    # Transform the list of dictionaries of dataframes into a dictionary of lists of
+    # dataframes, in which all dataframes in each list represent different instances of
+    # the same page of data from different years
+    all_data = defaultdict(list)
+    for dfs in paged_dfs:
+        for page in dfs:
+            all_data[page].append(dfs[page])
+
+    # concatenate the dataframes in each list in the dictionary into a single dataframe
+    for page in all_data:
+        all_data[page] = pd.concat(all_data[page]).reset_index(drop=True)
+
+    return all_data
+
+
+def year_extractor_factory(
+    extractor_cls: type[GenericExtractor], name: str
+) -> OpDefinition:
+    """Construct a Dagster op that extracts one year of data, given an extractor class.
+
+    Args:
+        extractor_cls: Class of type :class:`Extractor` used to extract the data.
+        name: Name of an Excel based dataset (e.g. "eia860").
+    """
+
+    def extract_single_year(context, year: int) -> dict[str, pd.DataFrame]:
```

Comment:
This is definitely a nitpick, but my preference when using factories like this to produce `op`s and `asset`s is to just apply the decorator to the internal functions like:

```py
def year_extractor_factory(
    extractor_cls: type[GenericExtractor], name: str
) -> OpDefinition:
    @op(
        required_resource_keys={"datastore", "dataset_settings"},
        name=f"extract_single_{name}_year",
    )
    def extract_single_year(context, year: int) -> dict[str, pd.DataFrame]:
        ...

    return extract_single_year